adam/visualization/make_scenes.py

```diff
@@ -22,46 +28,89 @@
     HasColor,
     HasBinaryProperty,
     ObjectPerception,
+    Relation,
 )
 from adam.ontology import OntologyNode
 
 from adam.visualization.panda3d_interface import SituationVisualizer
 from adam.visualization.utils import Shape
 
 
+@attrs(slots=True)
+class SceneNode:
+    elt: ObjectPerception = attr.ib()
+    children: List["SceneNode"] = attr.ib(factory=list)
+    parent: "SceneNode" = attr.ib(default=None)
+
+
 def main() -> None:
     random.seed(2015)
 
-    # go through curriculum scenes (fed in from where?) and output geometry types
+    # go through curriculum scenes and output geometry types
     print("scene generation test")
     viz = SituationVisualizer()
-    for i, scene in enumerate(SceneCreator.create_scenes(GAILA_PHASE_1_CURRICULUM)):
+    for i, (property_map, obj_graph) in enumerate(
+        SceneCreator.create_scenes(GAILA_PHASE_1_CURRICULUM)
+    ):
         print(f"SCENE {i}")
-        for obj, properties in scene.items():
-            print(f"Object: {obj}")
-            # only interested in rendering geons
-            # TODO***: allow for Irregular geons to be rendered
-            if (
-                obj.geon is None
-                or SceneCreator.cross_section_to_geo(obj.geon.cross_section)
-                == Shape.IRREGULAR
-            ):
-                continue
-            color = None
-            for prop in properties:
-                if isinstance(prop, RgbColorPerception):
-                    color = prop
-            viz.add_model(
-                SceneCreator.cross_section_to_geo(obj.geon.cross_section),
-                SceneCreator.random_position(),
-                color,
-            )
+        # for debugging purposes:
+        SceneCreator.graph_for_each(obj_graph, obj_names)
+
+        # bind visualizer and properties to render function:
+        bound_render_obj = partial(
+            render_obj, viz, property_map
+        )  # bind renderer to function
+        bound_render_nested_obj = partial(render_obj_nested, viz, property_map)
+        # render each object in graph
+        SceneCreator.graph_for_each_top_level(
+            obj_graph, bound_render_obj, bound_render_nested_obj
+        )
         viz.run_for_seconds(2.25)
+        viz.print_scene_graph()
         input("Press ENTER to continue")
         viz.clear_scene()
         viz.run_for_seconds(0.25)
 
 
+def render_obj(
+    renderer: SituationVisualizer,
+    properties: DefaultDict[
+        ObjectPerception, List[Optional[Union[RgbColorPerception, OntologyNode]]]
+    ],
+    obj: ObjectPerception,
+) -> NodePath:
+    return render_obj_nested(renderer, properties, obj, None)
+
+
+def render_obj_nested(
+    renderer: SituationVisualizer,
+    properties: DefaultDict[
+        ObjectPerception, List[Optional[Union[RgbColorPerception, OntologyNode]]]
+    ],
+    obj: ObjectPerception,
+    parent: NodePath,
```

Comment:
Shouldn't the type annotation here be `Optional[NodePath]` if we are passing `None` in `render_obj`